network/transport/v2/protocol.go

```diff
@@ -71,22 +116,135 @@ func (p protocol) UnwrapMessage(envelope interface{}) interface{} {
 func (p protocol) Configure(_ transport.PeerID) {
 }
 
-func (p protocol) Start() {
+func (p *protocol) setupNatsHandler() error {
+	conn, err := p.natsConnectHandler(
+		p.config.Nats.Hostname,
+		p.config.Nats.Port,
+		time.Duration(p.config.Nats.Timeout)*time.Second,
+	)
+	if err != nil {
+		return err
+	}
+
+	js, err := conn.JetStream()
+	if err != nil {
+		return err
+	}
+
+	if _, err = js.Subscribe(events.PrivateTransactionsSubject, func(msg *nats.Msg) {
+		if err := p.handlePrivateTx(msg); err != nil {
+			logrus.Errorf("failed to handle private transaction: %v", err)
+
+			if err := msg.Nak(); err != nil {
+				logrus.Errorf("failed to NACK private transaction event: %v", err)
+			}
+
+			return
+		}
+
+		if err := msg.Ack(); err != nil {
+			logrus.Errorf("failed to ACK private transaction event: %v", err)
+		}
+	}); err != nil {
+		conn.Close()
+		return err
+	}
+
+	go func() {
+		<-p.ctx.Done()
+		conn.Close()
+	}()
+
+	return nil
+}
+
+func (p *protocol) Start() {
+	p.ctx, p.cancel = context.WithCancel(context.Background())
+
+	go func() {
+		for {
+			if err := p.setupNatsHandler(); err != nil {
+				logrus.Errorf("failed to setup NATS handler: %v", err)
+
+				select {
+				case <-p.ctx.Done():
+					return
+				case <-time.After(5 * time.Second):
+					continue
+				}
+			}
+
+			return
+		}
+	}()
 }
 
-func (p protocol) Stop() {
+func (p *protocol) handlePrivateTx(msg *nats.Msg) error {
+	if p.nodeDID == nil {
+		return errors.New("node DID is not set")
+	}
+
+	doc, _, err := p.docResolver.Resolve(*p.nodeDID, nil)
+	if err != nil {
+		return err
+	}
+
+	keyAgreementIDs := make([]string, len(doc.KeyAgreement))
+
+	for i, keyAgreement := range doc.KeyAgreement {
+		keyAgreementIDs[i] = keyAgreement.ID.String()
+	}
+
+	tx, err := dag.ParseTransaction(msg.Data)
+	if err != nil {
+		return err
+	}
+
+	if len(tx.PAL()) == 0 {
+		return errors.New("PAL header is empty")
+	}
+
+	epal := dag.EncryptedPAL(tx.PAL())
+
+	pal, err := epal.Decrypt(keyAgreementIDs, p.decrypter)
+	if err != nil {
+		return err
+	}
+
+	// We weren't able to decrypt the PAL, so it wasn't meant for us
+	if pal == nil {
+		return nil
+	}
+
+	conn := p.connectionList.Get(grpc.ByConnected(), grpc.ByNodeDID(pal[0]))
+
+	if conn == nil {
+		return fmt.Errorf("no connection found (DID=%s)", pal[0])
+	}
+
+	return conn.Send(p, &Envelope{Message: &Envelope_TransactionPayloadQuery{
```

Comment:
this doesn't actually guarantee the payload will be retrieved (just that the message is put in the outbox). Do we have a system in place to retrieve it in case `Send()` succeeds, but we never receive the payload?

Reply:
Right. I'll think of something. We probably need this in more places anyways.

Reply:
That should also be part of the retry mechanism